src/sqlfluff/utils/functional/context.py

```diff
@@ -0,0 +1,54 @@
+"""Define FunctionalContext class."""
+
+from sqlfluff.utils.functional import Segments
+from sqlfluff.core.rules import RuleContext
+
+
+class FunctionalContext:
+    """RuleContext written in a "functional" style; simplifies writing rules."""
+
+    def __init__(self, context: RuleContext):
+        self.context = context
+
+    @property
+    def segment(self) -> "Segments":
+        """Returns a Segments object for context.segment."""
+        return Segments(
+            self.context.segment, templated_file=self.context.templated_file
+        )
+
+    @property
+    def parent_stack(self) -> "Segments":  # pragma: no cover
+        """Returns a Segments object for context.parent_stack."""
+        return Segments(
+            *self.context.parent_stack, templated_file=self.context.templated_file
+        )
+
+    @property
+    def siblings_pre(self) -> "Segments":  # pragma: no cover
+        """Returns a Segments object for context.siblings_pre."""
+        return Segments(
+            *self.context.siblings_pre, templated_file=self.context.templated_file
+        )
+
+    @property
+    def siblings_post(self) -> "Segments":  # pragma: no cover
+        """Returns a Segments object for context.siblings_post."""
+        return Segments(
+            *self.context.siblings_post, templated_file=self.context.templated_file
+        )
+
+    @property
+    def raw_stack(self) -> "Segments":
+        """Returns a Segments object for context.raw_stack."""
+        return Segments(
+            *self.context.raw_stack, templated_file=self.context.templated_file
+        )
+
+    @property
+    def raw_segments(self):
+        """Returns a Segments object for all the raw segments in the file."""
+        file_segment = self.context.parent_stack[0]
+        return Segments(
+            *file_segment.get_raw_segments(), templated_file=self.context.templated_file
+        )
```

Comment:
```suggestion
        )
 
def functional_eval(func: Callable[[rule_self: Any, context: FunctionalContext]]):
    def wrapper(self, context: RuleContext):
        fn_context = FunctionalContext(context)
        return func(self, fn_context)
    return wrapper
```

Reply:
I'm going to address this as a separate issue and moved the discussion here: https://github.com/sqlfluff/sqlfluff/issues/3745